ENH: MultiIndex.from_frame

ArtinSarraf opened the pull request:
- [x] closes #22420
- [x] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This pull request is to add the from_frame method of creating multiindexes. Along with this feature the helper method "squeeze" has been added for squeezing single level multiindexes to a standard index (analogous to df.squeeze).

Additionally the to_frame method was updated to guarantee that the order of the labels is preserved when converting a multiindex to a dataframe. Currently this cannot be guaranteed in Python 2.7 due to the use of a dictionary for creating the frame. With this change from_frame and to_frame are perfectly complementary.

Commit:
ENH - add from_frame method and accompanying squeeze method to multiindex

Commit:
ENH - guarentee that order of labels is preserved in multiindex to_frame method

Commit:
CLN - adhere to PEP8 line length

Commit:
CLN - remove trailing whitespace

jreback commented:
huh? pls show a real usecase for this

in the future pls open an issue for discussion first

ArtinSarraf commented:
Sorry should have tagged this issue opened ~2 months ago. The issue contains an example as well.

https://github.com/pandas-dev/pandas/issues/22420

ArtinSarraf commented:
Here are some more concrete examples of how these can be used as well.
Adding metadata.

    df = pd.DataFrame(np.random.randn(10, 6), columns=pd.MultiIndex.from_tuples([('NY', 'KJFK'), ('NY', 'KLGA'), ('CA', 'KLAX'), ('CA', 'KBUR'), ('AB', 'CYEG'), ('AB', 'CYYC')], names=['state', 'ICAO']))
    country_state_mapping = pd.DataFrame([['US', 'NY'], ['US', 'CA'], ['CDN', 'AB'], ['CDN', 'ON']], columns=['country', 'state'])
    meta = df.columns.to_frame(index=False)
    new_meta = meta.merge(country_state_mapping, on='state', how='left')[['country', 'state', 'ICAO']]
    df.columns = pd.MultiIndex.from_frame(new_meta)

Filtering

    # continuing from previous
    meta = df.columns.to_frame(index=False)
    new_meta = meta.loc[meta.state.isin(['CA', 'AB'])]
    df = df.reindex(columns=pd.MultiIndex.from_frame(new_meta))

jreback commented:
you can do the same thing if everything is in long form.

```
In [93]: (pd.merge(df.unstack().reset_index(), country_state_mapping, on='state', how='left')
    ...:    .set_index(['country', 'state', 'ICAO', 'level_2'])
    ...:    .sort_index()
    ...:    .unstack(level=[0,1,2])
    ...:    .sort_index(axis=1)
    ...: )
Out[93]: 
                0                                                  
country       CDN                  US                              
state          AB                  CA                  NY          
ICAO         CYEG      CYYC      KBUR      KLAX      KJFK      KLGA
level_2                                                            
0        0.368735  0.428429 -1.173573  0.256197  1.792880  0.414737
1        0.993952 -0.313126 -1.814687 -0.025160 -0.775963 -0.665919
2       -0.635185  2.063095  0.063650 -0.939815  0.219432  1.489314
3       -0.156358  0.201501  0.344305 -1.562817 -0.103385  0.482686
4        0.551272 -1.628747 -1.387776 -1.400052 -1.493635  2.323896
5       -1.487852  0.466714  1.195675  0.006298  0.277463  1.170253
6       -0.453253  1.850580  0.240695  1.375471  0.144580  0.012171
7       -0.079981 -1.139398  0.855885  2.687475 -1.374322  0.543441
8       -1.432287  0.100697 -0.554310 -0.306952  1.011671 -0.723025
9       -0.235468  0.893887  0.638946  0.916163  0.149475  1.121286
```

What you are proposing is wide from manipulation, which though a nice idea, and in this particular case makes sense. Is not generaly purpose, nor easy for people to do (and requires sorting guarantees).

So why should this be added?

ArtinSarraf commented:
One of the nice features of multiindexes is that they act as meta-dataframes for the associated numeric data, allowing a frame to be much more concise and memory efficient. Long form manipulations require multiple pivots (which can be time consuming) and can require significant extra memory. The relationship between multiindexes and a standard dataframe can also be demonstrated through this following example. One can substitute a multiindex (and I have seen this done before) with a separate metaframe like so:

    df = pd.DataFrame(np.random.randn(10, 6))
    meta = pd.DataFrame([('NY', 'KJFK'), ('NY', 'KLGA'), ('CA', 'KLAX'), ('CA', 'KBUR'), ('AB', 'CYEG'), ('AB', 'CYYC')], columns=['state', 'ICAO'])
    df.loc[:, meta.loc[meta.state.isin(['NY', 'AB'])].index]  # equivalent to above filtering example

I believe that this, along with the fact that to_frame already exists shows that from_frame is a reasonable and logical complementary method to be added as a way to instantiate multiindexes.

Regarding the sorting guarantees of to_frame, I think that this is something that any user would reasonably expect, considering the similarities between a mi and a df. Additionally, not only does it keep manipulations, that might involve using it, consistent, but all other related casting methods (e.g. tolist) ensure the sorting of the resultant type matches that of the original multiindex. At the very least I think adding a parameter to to_frame for preserving the original sorting would make sense. It would be very easy to implement on the existing implementation.

    # to_frame method of MultiIndex
    def to_frame(self, index=True, preserve_order=False):
        from pandas import DataFrame
        result = DataFrame({(name or level):
                            self._get_level_values(level)
                            for name, level in
                            zip(self.names, range(len(self.levels)))},
                           copy=False)
        if preserve_order:
            result = result.reindex(index=self)
        if index:
            result.index = self
        return result

jreback reviewed (changes requested):
so need some discussions

and of course would need much testing - having some examples would acceptance possibikity

ArtinSarraf commented:
I would be happy to contribute to any further discussions. Is there a more appropriate forum for that?

Also I will add tests to the PR this week.

Are there any types of examples in particular that would help move along the addition of this feature?

Thanks.

jreback commented:
> Are there any types of examples in particular that would help move along the addition of this feature?

some usecases, and showing how its possible now, but how the 'new' method makes it easeir.

ArtinSarraf commented:
Here is a demo to show the benefits of the from_frame method. Thanks for taking the time to look.

    # for convenience of this demo
    pd.MultiIndex.from_frame = staticmethod(lambda df: pd.MultiIndex.from_tuples(list(df.values), names=list(df)))

    # memory usage of below frame ~82000000
    #    multiindexes this large are not uncommon when dealing with things such as weather data for many cities
    #    with their associated meta location info (i.e. country, state, county, etc.)
    df = df = pd.DataFrame(
        np.random.randn(10000, 1024), 
        index=pd.date_range('19910905', periods=10000, name='date'), 
        columns=pd.MultiIndex.from_product(
            [list('abcd'), list('efgh'), list('ijkl'), list('mnop'), list('qrst')], 
            names=['L1', 'L2', 'L3', 'L4', 'L5']
        )
    )

    meta_mapping = pd.DataFrame([['x', 'a'], ['x', 'b'], ['y', 'c'], ['z', 'd']], columns=['L0', 'L1'])


# Adding additional meta data

    # Long Form
    # peak memory usage ~573440080  --  ~7x larger than original df
    # execution time: ~13.7 secs

    (
        df.unstack().reset_index().merge(meta_mapping, on='L1', how='left')
        .set_index(['L0', 'L1', 'L2', 'L3', 'L4', 'L5', 'date'])[0].unstack([0, 1, 2, 3, 4, 5])
    )


    # Wide Form
    # peak memory usage ~82000000  --  no difference from original df
    # execution time: ~0.271 sec  --  ~50x faster than long form
    # cleaner and does not require explicitly writing out positional arguments or axis 0 names
    # requires guaranteed ordering of to_frame - easy to guarantee/implement

    (
        df.T.set_index(pd.MultiIndex.from_frame(
            df.columns.to_frame(index=False)
            .merge(meta_mapping, on='L1', how='left')[['L0', 'L1', 'L2', 'L3', 'L4', 'L5']]
        )).T
    )

# Quick Prototyping

    # Often find it easier to prototype meta frames in excel. I then need to use that frame as a multiindex.
    # Assume the below snippet was copied from excel rather than pyperclip
    pyperclip.copy('''
    country   state   county        city       
    USA       CT      Fairfield     Greenwich  
    USA       CT      Fairfield     Stamford   
    USA       NY      Westchester   Bronxville 
    USA       NY      Bronx         Bronx      
    USA       NJ      Monmouth      Rumson     
    USA       NJ      Monmouth      Middletown 
    USA       NJ      Ocean         Ocean      
    Canada    ON      York          Toronto    
    ''')
    meta = pd.read_clipboard()
    df = pd.DataFrame(np.random.randn(10, 8), columns=pd.MultiIndex.from_frame(meta))

    # As far as I know there is no good and quick way to do this currently. (aside from the from_tuples method I showed above)


# Complex Filtering/Slicing - All the following examples are related to filtering

    meta = df.columns.to_frame(index=False)  # used for all from_frame examples below


# Keep only counties with > 1 city

    # With from_frame
    df.reindex(columns=pd.MultiIndex.from_frame(meta.groupby('county').filter(lambda df: len(df)>1)))

    # Long Form - same disadvantages as stated above
    (
        df.unstack().reset_index().groupby('county').filter(lambda dfx: len(dfx.city.unique())>1)
        .set_index(['country', 'state', 'county', 'city', 'level_4']).unstack([0,1,2,3])[0]
    )


# Admittedly the 2 examples below can be done without from_frame as well, given to_frame is ordered, like so:

    mask = (meta...) & (meta...) | (meta...)
    df.loc[:, mask]


# Get CT + NJ without Greenwich

    # With from_frame
    filtered_df = meta.loc[(meta.state.isin(['CT', 'NJ'])) & (meta.city != 'Greenwich')]
    df.reindex(columns=pd.MultiIndex.from_frame(filtered_df))

    # Without from_frame
    # While this is more concise, it requires two different chained methods, while the from_frame
    # method simplifies the filter to a standard dataframe filter
    df.loc[:, pd.IndexSlice['USA', ['CT', 'NJ'], :]].drop('Greenwich', axis=1, level='city')


# Get all cities that don't start with "Bron"

    # With from_frame
    filtered_df = meta.loc[~meta.city.str.startswith('Bron')]
    df.reindex(columns=pd.MultiIndex.from_frame(filtered_df))

    # Without from_frame
    # Not sure of a good way without going to long form

Commit:
ENH - raise TypeError on inappropriate input

Commit:
TST - add tests for mi.from_frame and mi.squeeze

Commit:
CLN - pep8 adherence in tests

Commit:
CLN - last missed pep8 fix

mitstake committed:
BUG - remove pd.DataFrame in favor of local import

ArtinSarraf commented:
Is there a way to rerun the Azure pipeline tests? The tests only failed for “Windows py36_np14” environment and the failed tests don’t seem to be related to my commits at all. I’m guessing it was a blip in the testing framework.

Commit:
DOC - add more detailed docstrings for from_frame and squeeze

ArtinSarraf commented:
Please see edits to demo:
https://github.com/pandas-dev/pandas/pull/23141#issuecomment-430083542

Also any ideas on the tests? I don't see anything in the logs that point to any changes I've made.

ArtinSarraf commented:
@jreback

Please see additions to demo (Formatting and Custom Sorting):
https://github.com/pandas-dev/pandas/pull/23141#issuecomment-430083542

Also any ideas on the tests? I don't see anything in the logs that point to any changes I've made.

ArtinSarraf commented:
@jreback. Any thoughts on the examples provided above?

jreback commented:
@pandas-dev/pandas-core if anyone has thoughts here

datapythonista commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1412,45 @@ def from_product(cls, iterables, sortorder=None, names=None):
+        Parameters
+        ----------
+        df : pd.DataFrame
+            DataFrame to be converted to MultiIndex
```
Can you finish with a period.

datapythonista commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1412,45 @@ def from_product(cls, iterables, sortorder=None, names=None):
+        ----------
+        df : pd.DataFrame
+            DataFrame to be converted to MultiIndex
+        squeeze : bool
```
Can you add the default.

datapythonista commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1412,45 @@ def from_product(cls, iterables, sortorder=None, names=None):
+    @classmethod
+    def from_frame(cls, df, squeeze=True):
+        """
+        Make a MultiIndex from a dataframe
```
Finish with period. Run `./scripts/validate_docstrings.py pandas.MultiIndex.from_frame` to make sure the docstring follows all the standards we validate.

Use `DataFrame` instead of dataframe.

datapythonista commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1412,45 @@ def from_product(cls, iterables, sortorder=None, names=None):
+
+        Returns
+        -------
+        index : MultiIndex
```
no need for giving a name, just leave the type. Add a description in the next line (indented).

datapythonista commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1412,45 @@ def from_product(cls, iterables, sortorder=None, names=None):
+        --------
+        >>> df = pd.DataFrame([[0, u'green'], [0, u'purple'], [1, u'green'],
+                               [1, u'purple'], [2, u'green'], [2, u'purple']],
+                              columns=[u'number', u'color'])
```
this is missing the `...` for multine commands. Also  show the content of `df` after creating it.

datapythonista commented on pandas/core/indexes/multi.py:
```diff
@@ -1474,6 +1513,28 @@ def _sort_levels_monotonic(self):
+        """
+        Squeeze a single level multiindex to be a regular Index instane. If
+        the MultiIndex is more than a single level, return a copy of the
+        MultiIndex.
```
Can you run `validate_docstrings.py` for this docstring, and also read the contributing docstring documentation too. There are several issues.

datapythonista reviewed (changes requested):
Some issues with the docstrings. Also  would be nice to add `See Also` sections. If you can take a look at the documentation for the docstrings that would be great.

Commit:
DOC - update MultiIndex.from_frame and squeeze doctests to comply with the Pandas docstring guide

ArtinSarraf commented:
@datapythonista. Changes have been made and pushed. I've also added from_frame to the "See Also" of the other 3 contructor methods. Thanks.

gfyoung commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1415,61 @@ def from_product(cls, iterables, sortorder=None, names=None):
+        df : pd.DataFrame
+            DataFrame to be converted to MultiIndex.
+        squeeze : bool, default True
+            If df is a single column, squeeze multiindex to be a regular index.
```
multiiindex --> MultiIndex

gfyoung commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1415,61 @@ def from_product(cls, iterables, sortorder=None, names=None):
+        if squeeze:
+            return mi.squeeze()
+        else:
+            return mi
```
Personal nit: change the four lines of `squeeze` logic to one:

~~~
return mi.squeeze() if squeeze else mi
~~~

gfyoung commented on pandas/core/indexes/multi.py:
```diff
@@ -1474,6 +1532,27 @@ def _sort_levels_monotonic(self):
 
+    def squeeze(self):
+        """
+        Squeeze a single level MultiIndex to be a regular Index instane.
```
instane --> instance

Commit:
CLN - cleanup docstrings and source

Commit:
TST - reorganize some of the multiindex tests

Commit:
CLN - adhere to pep8 line length

ArtinSarraf commented:
@gfyoung. All requested changes have been applied. Thanks.

jreback commented:
before we get more comments on this PR code itself. would like commentary on the merits of this idea. 

@ArtinSarraf actually might be beneficial to open an issue about this first.

jreback commented:
@ArtinSarraf i see you did actually.

ArtinSarraf commented:
@jreback - I’ve referenced a couple of the important comments in this PR in the associated issue.

TomAugspurger commented on pandas/core/indexes/multi.py:
```diff
@@ -1189,11 +1189,11 @@ def to_frame(self, index=True, name=None):
-                            for name, level in
-                            zip(idx_names, range(len(self.levels)))},
-                           copy=False)
+        result = DataFrame(
```
Doesn't this lose CategoricalDtype info?

Can you ensure we have a test for

```python
In [12]: idx = pd.MultiIndex.from_product([['a', 'b'], pd.Categorical(['A', 'B'])])

In [13]: idx.to_frame().dtypes
Out[13]:
0      object
1    category
dtype: object

```

and make sure that doesn't regress?

ArtinSarraf replied:
Any reason not to just use OrderedDict here? Would satisfy both the guaranteed ordering and dtype fidelity.

TomAugspurger replied:
Seems reasonable to me.

TomAugspurger commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1415,58 @@ def from_product(cls, iterables, sortorder=None, names=None):
+        if not isinstance(df, DataFrame):
+            raise TypeError("Input must be a DataFrame")
+        columns = list(df)
+        mi = cls.from_tuples(list(df.values), names=columns)
```
Again, I would expect that the dtypes are preserved here.

ArtinSarraf replied:
Will try out something along these lines when I’m at a computer.

`pd.MultiIndex.from_arrays([df[x] for x in df])`

Commit:
BUG - ensure dtypes are preserved in from_frame and to_frame

Commit:
TST - add tests for ensuring dtype fidelity and custom names for from_frame

ArtinSarraf commented:
@TomAugspurger - I've made the changes to ensure that the dtypes are preserved in both from/to_frame. I've also added tests to both to catch this. Also added one extra ability to from_frame, passing an optional names argument for overriding the default names (similar to the to_frame method).

Commit:
CLN - pep8 adherence

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1420,77 @@ def from_product(cls, iterables, sortorder=None, names=None):
+    def from_frame(cls, df, squeeze=True, names=None):
+        """
+        Make a MultiIndex from a DataFrame.
+
```
add a versionadded tag

ArtinSarraf replied:
Found the version from the pending what's new page. Added the tag for 0.24.0, and updated the what's new page.

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1420,77 @@ def from_product(cls, iterables, sortorder=None, names=None):
+        ----------
+        df : pd.DataFrame
+            DataFrame to be converted to MultiIndex.
+        squeeze : bool, default True
```
should be the last kwarg as much less common

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1420,77 @@ def from_product(cls, iterables, sortorder=None, names=None):
+            DataFrame to be converted to MultiIndex.
+        squeeze : bool, default True
+            If df is a single column, squeeze MultiIndex to be a regular Index.
+        names : list / sequence / callable, optonal
```
optional

use list-like

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1420,77 @@ def from_product(cls, iterables, sortorder=None, names=None):
+        names : list / sequence / callable, optonal
+            If no names provided, use column names, or tuple of column names if
+            the columns is a MultiIndex. If sequence, overwrite names with the
+            given sequence. If callable, pass each column name or tuples of
```
not really sure of the difference of these, can you show what the rationale for all of these options?

ArtinSarraf replied:
The callable option was mostly for cases where the frame used to construct the mi is itself multiindexed on the columns. Example below:

        >>> df = pd.DataFrame([['a', 'b'], ['c', 'd'], ['e', 'f']],
        ...                   columns=pd.MultiIndex.from_tuples([('L1', 'x'), 
        ...                                                      ('L2', 'y')]))
        >>> df
          L1 L2
           x  y
        0  a  b
        1  c  d
        2  e  f
        >>> pd.MultiIndex.from_frame(df, names=lambda x: '_'.join(x))
        MultiIndex(levels=[['a', 'c', 'e'], ['b', 'd', 'f']],
                   labels=[[0, 1, 2], [0, 1, 2]],
                   names=['L1_x', 'L2_y'])

ArtinSarraf replied:
This would be an uncommon occurrence. Would it make more sense to just not provide the callable option?

jreback replied:
let's remove the callable option for now

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1420,77 @@ def from_product(cls, iterables, sortorder=None, names=None):
+        5       2    joy
+        >>> pd.MultiIndex.from_frame(df)
+        MultiIndex(levels=[[0, 1, 2], ['happy', 'jolly', 'joy']],
+                   labels=[[0, 0, 1, 1, 2, 2], [0, 1, 0, 1, 2, 2]],
```
can you show examples using the other options

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -1474,6 +1553,30 @@ def _sort_levels_monotonic(self):
+    def squeeze(self):
+        """
+        Squeeze a single level MultiIndex to be a regular Index instance.
+
```
I don't think we need to make this a public method.

ArtinSarraf replied:
The rationale for making this a public method was, in bigger projects I will sometimes receive dataframes that I have no control over. Sometimes the columns would be formatted as a single level multiindex (I think this was mostly due to processes setup in earlier versions of pandas where the index would otherwise have been converted automatically to an Index). The easiest way to deal with these cases was to "squeeze" the index.
It's not a common occurrence though. Is this too obscure to justify keeping it public?

Also the context that it is used here is in a classmethod, so it's being called on the MultiIndex object externally. Would it be ok to call a "private" method in this fashion?

ArtinSarraf replied:
changed squeeze -> _squeeze

jreback replied:
I am re-thinking if this should be public, see #22866

jreback commented on pandas/tests/indexes/multi/test_constructor.py:
```diff
@@ -472,3 +472,77 @@ def test_from_tuples_with_tuple_label():
+        'b': pd.Categorical(['a', 'a', 'b', 'b', 'c', 'c'], ordered=True),
+        'c': ['x', 'x', 'y', 'z', 'x', 'y']
+    })
+    original_dtypes = df.dtypes.to_dict()
```
construct an expected frame and use ``tm.assert_frame_equal``

jreback commented on pandas/tests/indexes/multi/test_constructor.py:
```diff
@@ -472,3 +472,77 @@ def test_from_tuples_with_tuple_label():
+    mi_dtypes = {name: mi.levels[i].dtype for i, name in enumerate(mi.names)}
+    assert original_dtypes == mi_dtypes
+
+
```
ideally you can parametrize these tests on names

jreback commented on pandas/tests/indexes/multi/test_conversion.py:
```diff
@@ -82,6 +82,20 @@ def test_to_frame():
+        pd.Categorical(['a', 'a', 'b', 'b', 'c', 'c'], ordered=True),
+        ['x', 'x', 'y', 'z', 'x', 'y']
+    ], names=['dates', 'a', 'b', 'c'])
+    original_dtypes = {name: mi.levels[i].dtype 
```
also construct an expected frame

Commit:
DOC - add examples and change order of kwargs for from_frame

Commit:
TST - parameterize tests

Commit:
CLN - pep8 adherence

Commit:
CLN - pep8 adherence

ArtinSarraf commented:
Since I have a lot of CLN commits to fix pep8 issues, would it be recommended to squash them all, or should I just leave them?

Commit:
DOC/CLN - add versionadded tags, add to whatsnew page, and clean up if-else logic branches to be simpler to follow

Commit:
CLN - squeeze -> _squeeze

mitstake committed:
DOC - squeeze -> _squeeze in whatsnew

ArtinSarraf commented:
@pandas-dev/pandas-core

Anymore feedback on this PR?

ArtinSarraf commented:
@jreback / @TomAugspurger / @gfyoung 

Any other feedback on code/merits of this PR?

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -892,6 +892,9 @@ MultiIndex
 
 - Removed compatibility for :class:`MultiIndex` pickles prior to version 0.8.0; compatibility with :class:`MultiIndex` pickles from version 0.13 forward is maintained (:issue:`21654`)
 - :meth:`MultiIndex.get_loc_level` (and as a consequence, ``.loc`` on a :class:`MultiIndex`ed object) will now raise a ``KeyError``, rather than returning an empty ``slice``, if asked a label which is present in the ``levels`` but is unused (:issue:`22221`)
+- :meth:`MultiIndex.from_frame` added. Allows constructing a `MultiIndex` object from a `DataFrame` (:issue:`22420`)
```
can you say ``MultiIndex`` has gain the ``.from_frame`` method, it allows .....

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -892,6 +892,9 @@ MultiIndex
 - Removed compatibility for :class:`MultiIndex` pickles prior to version 0.8.0; compatibility with :class:`MultiIndex` pickles from version 0.13 forward is maintained (:issue:`21654`)
 - :meth:`MultiIndex.get_loc_level` (and as a consequence, ``.loc`` on a :class:`MultiIndex`ed object) will now raise a ``KeyError``, rather than returning an empty ``slice``, if asked a label which is present in the ``levels`` but is unused (:issue:`22221`)
+- :meth:`MultiIndex.from_frame` added. Allows constructing a `MultiIndex` object from a `DataFrame` (:issue:`22420`)
+- :meth:`MultiIndex._squeeze` added. Allows a `MultiIndex` with only a single level to be converted to an `Index` object (:issue:`22420`)
```
this is not public, remove the note

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -892,6 +892,9 @@ MultiIndex
 - :meth:`MultiIndex.get_loc_level` (and as a consequence, ``.loc`` on a :class:`MultiIndex`ed object) will now raise a ``KeyError``, rather than returning an empty ``slice``, if asked a label which is present in the ``levels`` but is unused (:issue:`22221`)
+- :meth:`MultiIndex.from_frame` added. Allows constructing a `MultiIndex` object from a `DataFrame` (:issue:`22420`)
+- :meth:`MultiIndex._squeeze` added. Allows a `MultiIndex` with only a single level to be converted to an `Index` object (:issue:`22420`)
+- :meth:`MultiIndex.to_frame` will now guarantee the preservation of the sort order of the level names in the resulting `DataFrame` (:issue:`22420`)
```
this whatsnew note move to api breaking changes

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -1189,11 +1190,15 @@ def to_frame(self, index=True, name=None):
         else:
             idx_names = self.names
 
```
can you add a comment here guarantiing sort order

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1420,107 @@ def from_product(cls, iterables, sortorder=None, names=None):
+        ----------
+        df : pd.DataFrame
+            DataFrame to be converted to MultiIndex.
+        sortorder : int or None
```
move sortorder to the end of the arg list (before or after squeeze is ok)

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1420,107 @@ def from_product(cls, iterables, sortorder=None, names=None):
+            Level of sortedness (must be lexicographically sorted by that
+            level).
+        names : list-like / callable, optonal
+            If no names provided, use column names, or tuple of column names if
```
optional

If no names are provided, use the column names, or .....

If a squence, overwrite....

If a callable

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1420,107 @@ def from_product(cls, iterables, sortorder=None, names=None):
+           number
+        0       0
+        1       1
+        >>> pd.MultiIndex.from_frame(df)
```
can you put a blank line between cases

jreback replied:
adding a comment on the case if its not obvious

ArtinSarraf replied:
removed other examples as I have removed the callable names /squeeze parameter

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1420,107 @@ def from_product(cls, iterables, sortorder=None, names=None):
+        MultiIndex.from_product : Make a MultiIndex from cartesian product
+                                  of iterables
+        """
+        from pandas import DataFrame
```
use ABCDataFrame

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1420,107 @@ def from_product(cls, iterables, sortorder=None, names=None):
+
+        # Get MultiIndex names
+        if names is None:
+            names = list(df)
```
df.columns ?

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1420,107 @@ def from_product(cls, iterables, sortorder=None, names=None):
+        elif callable(names):
+            names = [names(x) for x in list(df)]
+        elif is_list_like(names):
+            if len(names) != len(list(df)):
```
use df.columns

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1420,107 @@ def from_product(cls, iterables, sortorder=None, names=None):
+            if len(names) != len(list(df)):
+                    raise ValueError("'names' should have same length as "
+                                     "number of columns in df.")
+            # else: use the passed in sequence
```
what is this comment?

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1420,107 @@ def from_product(cls, iterables, sortorder=None, names=None):
+                            "names, or a callable.")
+
+        # This way will preserve dtype of columns
+        mi = cls.from_arrays([df[x] for x in df],
```
I would raise on ``not df.columns.is_unique``

ArtinSarraf replied:
since non-unique names are valid in multiindex I have updated from_frame to provide the analogous multiindex even in the case of repeat names (like so: `[df.iloc[:, x] for x in range(len(df.columns))]`

jreback commented on pandas/tests/indexes/multi/test_constructor.py:
```diff
@@ -472,3 +472,83 @@ def test_from_tuples_with_tuple_label():
     idx = pd.MultiIndex.from_tuples([(2, 1), (4, (1, 2))], names=('a', 'b'))
     result = pd.DataFrame([2, 3], columns=['c'], index=idx)
     tm.assert_frame_equal(expected, result)
+
```
can you add something like before this section (and do the same for the other big sections)

```
    # ------------------------------------------------------------------------
    # from_frame
    # ------------------------------------------------------------------------
 ```

jreback commented on pandas/tests/indexes/multi/test_constructor.py:
```diff
@@ -472,3 +472,83 @@ def test_from_tuples_with_tuple_label():
+    tm.assert_index_equal(expected, result)
+
+
+def test_from_frame_non_frame():
```
can you parametrize over some common input types (2-d np array, Index, Series, scalar)

jreback commented on pandas/tests/indexes/multi/test_constructor.py:
```diff
@@ -472,3 +472,83 @@ def test_from_tuples_with_tuple_label():
+    assert original_dtypes == mi_dtypes
+
+
+@pytest.mark.parametrize('names_in,names_out', [
```
where is the test about guaranteed sort order?

ArtinSarraf replied:
Added, this also raised a minor bug in to_frame. If you have a column name 0 in your columns not in the 0th position, the name will be replaced with its positional index. I have updated to_frame to account for this.

ArtinSarraf replied:
consider these 2 examples:

    >>> pd.MultiIndex.from_arrays([[1,2,3],[1,2,3],[1,2,3]], names=[2,1,0]).to_frame()
    >>> pd.MultiIndex.from_arrays([[1,2,3],[1,2,3],[1,2,3]], names=['a',0,'b']).to_frame()

This is due to the `(name or level)` key in the dict comprehension.
I'd be surprised if anyone ever ran into this, but the update to to_frame handles this now.

jreback commented on pandas/tests/indexes/multi/test_conversion.py:
```diff
@@ -82,6 +82,29 @@ def test_to_frame():
     tm.assert_frame_equal(result, expected)
 
 
+def test_to_frame_dtype_fidelity():
```
can you add the issue number (ideally add to most of the tests you added, where appropriate)

jreback reviewed (changes requested):
looking good.

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -1412,6 +1420,107 @@ def from_product(cls, iterables, sortorder=None, names=None):
+            the columns is a MultiIndex. If sequence, overwrite names with the
+            given sequence. If callable, pass each column name or tuples of
+            names to the callable.
+        squeeze : bool, default True
```
yeah was thinking of that as well. Maybe we should expose ``.squeeze()`` as ``.to_index()`` (or  ``.to_flat_index()``, though likeing squeeze more and more

ArtinSarraf replied:
Have removed squeeze parameter (and method in accordance with @toobaz 's comment). Can't imagine this would be needed often at all.

toobaz commented:
I'm not convinced ``squeeze`` should be exposed - not even as a parameter. ``._squeeze()`` is - in the only case in which it makes sense - just an alternative for ``.get_level_values(0)``. And in general in the API we do not distinguish much between ``Index`` and ``MultiIndex`` with 1 level. In fact, I guess any difference (not just in behavior, also in performance) can be considered a bug.

So why should (unsophisticated) users bother about this difference?!

toobaz commented:
Another thought: is there any reason why we should not accept a 2-d ``array``, or even list of lists, rather than just a ``DataFrame``)? After all, we are primarily looking at the ``df.values`` (although ``.columns`` is useful for the level names). We would have to change the method name (``from_data``?), but we would mirror slightly better both the initialization of a ``pd.Series(a_1d_array)``, and the interchangeability of 2d structures (``DataFrame``/``MultiIndex``/``array``/nested lists) as right hand side of an assignment.

ArtinSarraf commented:
@toobaz - Agreed regarding the squeeze method. The original inspiration for it was dealing with dataframes coming from external processes that I have no control over which contained single level multiindexes. I think almost all of these cases were due to bad coding practices that in older versions of pandas would have resulted in a regular index instead (e.g. df.columns = [['a','b','c']]). The only time I would see it having any usefulness over get_level_values(0) is if you don't know whether an object is a multilevel multiindex or not, which probably shouldn't ever be the case. That certainly isn't the case here, so I really should just be using get_level values, and will update accordingly.

> And in general in the API we do not distinguish much between Index and MultiIndex with 1 level. In fact, I guess any difference (not just in behavior, also in performance) can be considered a bug.

Off the top of my head, groupby, while behaving as I would expect, is slightly different.

df.groupby('col1') -> df.groupby(('col1',))

jreback commented:
only problem with accepting generic 2d input is we lose the names
but that would be ok if names is None

so agree this would be a nice addition (just need to wrap the input in the Dataframe constructor)

ArtinSarraf commented:
@toobaz regarding the second commnet, would that maybe make more sense as an orient parameter in from_arrays?

Commit:
BUG - allow repeat column names in from_frame, and falsey column names in to_frame

Commit:
DOC - whatsnew formatting

Commit:
TST - reorganize and add tests for more incompatible from_frame types

Commit:
Merge branch 'enhancement/from_frame' of https://github.com/ArtinSarraf/pandas into enhancement/from_frame

Commit:
CLN - remove squeeze tests

Commit:
CLN - remove squeeze parameter from from_frame

Commit:
Merge branch 'master' into enhancement/from_frame

Commit:
TST - remove callable name option

toobaz commented:
> @toobaz regarding the second commnet, would that maybe make more sense as an orient parameter in from_arrays?

Good point... but then, I would as well merge the functionality of ``from_frame`` in ``from_arrays``.

Not just for avoiding duplication in the API, but also because I find it wrong (I had never noticed) that

```
In [2]: d = [[1, 2], [3, 4]]

In [3]: pd.MultiIndex.from_arrays(pd.DataFrame(d))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
[...]
TypeError: Input must be list-like

In [4]: pd.MultiIndex.from_arrays(np.array(d))
Out[4]: 
MultiIndex(levels=[[1, 2], [3, 4]],
           labels=[[0, 1], [0, 1]])
```

Again: my point is that, where possible, we treat all kinds of 2d data in the same way. So if ``from_arrays`` is able to manage a 2d array, it should also be able to manage a ``pd.DataFrame``. Now, as you correctly point out, the orientation changes - and it would be cool to have ``from_arrays`` to manage the different orientation, rather than a new method. I would just not add a new method only for taking care of column names, which are optional.

ArtinSarraf commented:
If from_arrays is updated to take a DataFrame, it will need to have a very specific implementation for them. First it will need to implicitly change the default orientation. Even if the default is set to None so that it can be inferred, it would be strange that the from_arrays(df) and from_arrays(df.values) would be different (since df.values probably shouldn’t be treated differently than any other 2d list).

Also from_frame is not just a replacement for from_arrays(df.values.T) since that would potentially lose dtype info (e.g. categoricals).

I’m not opposed to moving the functionality into somewhere more generic, but I think tailoring from_arrays to allow DataFrames might introduce quite a bit more inconsistencies and specificity tailored logic than we would want in a generic method.

toobaz commented:
> If from_arrays is updated to take a DataFrame, it will need to have a very specific implementation for them. First it will need to implicitly change the default orientation. Even if the default is set to None so that it can be inferred, it would be strange that the from_arrays(df) and from_arrays(df.values) would be different (since df.values probably shouldn’t be treated differently than any other 2d list).

What I had in mind would actually be that ``from_arrays(df)`` would behave exactly like ``from_arrays(df.values)`` and that ``from_arrays(df, orient="horizontal")`` would behave like you are designing ``from_frame(df)``. But I do realize now that one expects a ``DataFrame`` to be transformed (by default) in a ``MultiIndex`` where each column becomes a level. So my proposal is not good.

I think the original sin is in ``from_arrays`` which should have defaulted to not changing the orientation of data, and I assume it is too late to change this. So you are right that ``from_arrays(df)`` is not the correct replacement for ``from_frame``. But I still believe that:
- the difference between the new method and ``from_arrays`` should primarily be of orientation, not of type: it should accept any 2d data structure (and ``from_arrays`` too, but now that's a separate issue)
- the name should reflect this. I would prefer ``from_data`` or even better ``from_array`` to ``from_frame``

TomAugspurger commented:
Agreed that the orientation difference warrants two methods.

________________________________
From: ArtinSarraf <notifications@github.com>
Sent: Monday, November 12, 2018 12:36:17 PM
To: pandas-dev/pandas
Cc: Tom Augspurger; Mention
Subject: Re: [pandas-dev/pandas] Enhancement/from frame (#23141)


If from_arrays is updated to take a DataFrame, it will need to have a very specific implementation for them. First it will need to implicitly change the default orientation. Even if the default is set to None so that it can be inferred, it would be strange that the from_arrays(df) and from_arrays(df.values) would be different (since df.values probably shouldn’t be treated differently than any other 2d list).

Also from_frame is not just a replacement for from_arrays(df.values.T) since that would potentially lose dtype info (e.g. categoricals).

I’m not opposed to moving the functionality into somewhere more generic, but I think tailoring from_arrays to allow DataFrames might introduce quite a bit more inconsistencies and specificity tailored logic than we would want in a generic method.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/pandas-dev/pandas/pull/23141#issuecomment-437985400>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABQHIpmVdqpACWOsJCo0wnKevRgs91G4ks5uub-hgaJpZM4Xa2wj>.

toobaz commented:
> Agreed that the orientation difference warrants two methods.

By the way: ``from_tuples`` does follow the other orientation... it is too specific both in the name and implementation, but if we make this new method general (e.g. ``from_data``) then we will be able to eventually deprecate ``from_tuples``.

Commit:
ENH - from_data initial commit

ArtinSarraf commented:
@toobaz / @TomAugspurger / @jreback - I just pushed a concept for `from_data`. It incorporates the orient parameter, and implements @jreback 's idea of wrapping inputs in the DataFrame constructor. This allows all inputs to be handled in a single consistent way.

TomAugspurger commented:
I'm not wild about the name `from_data` :/

I'd almost prefer keeping from_dataframe. We could choose to wrap
non-dataframes inputs
in the dataframe constructor or not. But to me the clarity of the name and
purpose of from_dataframe
outweighs any loss of generality.

On Tue, Nov 13, 2018 at 9:19 PM ArtinSarraf <notifications@github.com>
wrote:

> @toobaz <https://github.com/toobaz> / @TomAugspurger
> <https://github.com/TomAugspurger> / @jreback <https://github.com/jreback>
> - I just pushed a concept for from_data. It incorporates the orient
> parameter, and implements @jreback <https://github.com/jreback> 's idea
> of wrapping inputs in the DataFrame constructor. This allows all inputs to
> be handled in a single consistent way.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/23141#issuecomment-438523431>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIvl1o2W2hudRsF9ESz7uYnxben1mks5uu4uqgaJpZM4Xa2wj>
> .
>

toobaz commented:
> I'm not wild about the name `from_data` :/

- ``from_array``? (yes, similar to ``from_arrays``, but at least makes it obvious to contrast them and get the difference)
- ``from_2d``?
- ``from_labels``?

``from_frame`` would be, I think, misleading and not in line with the other methods (e.g. ``from_arrays``, not ``from_Series``)...

toobaz commented:
> It incorporates the orient parameter,

Sorry, I'm confused: what does it add to ``from_arrays``?

ArtinSarraf commented:
@toobaz / @TomAugspurger - any thoughts on the consolidated from_data implementation?

toobaz commented on pandas/core/indexes/multi.py:
```diff
@@ -1417,6 +1427,103 @@ def from_product(cls, iterables, sortorder=None, names=None):
+            df = df.T
+
+        if not is_df:
+            df.columns = [None for _ in range(len(df.columns))]
```
Do we really need this? I think having levels numbered (like default df columns) is better.

toobaz commented on pandas/core/indexes/multi.py:
```diff
@@ -1417,6 +1427,103 @@ def from_product(cls, iterables, sortorder=None, names=None):
+                            "names.")
+        return cls.from_arrays([df.iloc[:, x] for x in range(len(df.columns))],
+                               sortorder=sortorder,
+                               names=names)
```
We are potentially creating the df twice, right? (here, and at 1505) I don't think it's good

toobaz commented on pandas/core/indexes/multi.py:
```diff
@@ -1417,6 +1427,103 @@ def from_product(cls, iterables, sortorder=None, names=None):
+                                     "number of columns in df.")
+        else:
+            raise TypeError("'names' must be a list / sequence of column "
+                            "names.")
```
This stuff will be checked by the constructor, I think this check is superfluous

toobaz commented on pandas/core/indexes/multi.py:
```diff
@@ -1417,6 +1427,103 @@ def from_product(cls, iterables, sortorder=None, names=None):
+        except ValueError:
+            raise TypeError("'from_data' input must be valid DataFrame input.")
+        if orient == 'rows':
+            df = df.T
```
I would personally prefer a ``rotate=True|False`` parameter. Not a very strong preference, but it is not easy for me to interpret the values of ``orient``.

toobaz reviewed (changes requested):
some coments

toobaz commented:
> any thoughts on the consolidated from_data implementation?

I left some comments.

In any case, I love the idea of ``from_data``, but because it allows to replace a lot of other methods (at least in the code). I definitely don't like adding ''both'' ``from_data`` ''and'' ``from_frame``. Even less if ``from_frame`` is ''not'' implemented through ``from_data``.

ArtinSarraf commented:
@toobaz - Definitely just one or the other. Seems like the consensus is leaning towards keeping from_frame for now.

toobaz commented:
> @toobaz - Definitely just one or the other. Although, seems like the consensus is leaning towards keeping from_frame for now.

@TomAugspurger expressed a preference (I think) over the name, not over the actual behavior/arguments. And I proposed some alternatives for the name (because I don't see the point in calling ``from_frame`` something which works on any 2D). What do you think of them?

ArtinSarraf commented:
@toobaz - I did misinterpret @TomAugspurger comment, but I was also considering [this review comment](https://github.com/pandas-dev/pandas/pull/23141#discussion_r234455896) by @jreback. Although this comment could have just been for the purposes of keeping the code true to the PR?

First, I like the idea of the generalized method that takes any 2d data including a df. However, since it does cover the cases of from_arrays and from_tuples, I would think it would be redundant and thus, either:
1) Have the single general method and plan to deprecate the redundant methods
2) Keep existing methods + from_frame (and only allow DataFrame input)

Regarding the name, if we go with option 1, I like the name from_data, since I think from_[data]frame misses a lot of the scope of the method. The only other names I can think of that could possibly adequately capture the scope of the method are "from_2d" or "from_table".

jreback commented:
> Have the single general method and plan to deprecate the redundant methods

not likely to deprecate these methods, they are very highly used


> Keep existing methods + from_frame (and only allow DataFrame input)

this is fine with me. I don't think the 2-D case is actually that interesting except for frames.

Commit:
DOC - reduce whatsnew entry for to_frame

Commit:
CLN/DOC - add examples to from_frame docstring and make code more readable

Commit:
Merge branch 'master' into enhancement/from_frame

Commit:
TST - use OrderedDict for dataframe construction

toobaz commented:
> * Have the single general method and plan to deprecate the redundant methods
> 
> * Keep existing methods + from_frame (and only allow DataFrame input)

I don't see these as the only alternatives. We can certainly add ``from_data`` and leave ``from_tuple`` and ``from_arrays`` as they are (internally implemented as thin wrappers around ``from_data``). And sure, we can also add ``from_frame`` as another minimal wrapper. I'm not in love with it, but it would need very little code. However I do think that ``from_data`` would be a clean and useful method: once you learn it, you can use it to create ``MultiIndex``es from _any_ kind of 2D data, with a common interface.

ArtinSarraf commented:
@toobaz - I could see this working well either way and like the from_data idea. However, in the interest of not getting stuck on what to do here, I would go with @jreback and @TomAugspurger choice and have just the dedicated from_frame method for now. And revisit from_data once the ability to instantiate a MultiIndex from a DataFrame has been established.

Commit:
Merge branch 'master' into enhancement/from_frame

ArtinSarraf commented:
@jreback / @TomAugspurger / @toobaz - anything else to consider to move this PR forward? Also any ideas on why some of the Tracis CI tests fail? The only thing I can think of is the change in the docstring section order causing validation to fail

TomAugspurger commented:
Sorry for the delay.

There are a few lining issues: https://travis-ci.org/pandas-dev/pandas/jobs/460592619#L2897

```
Check import format using isort
ERROR: /home/travis/build/pandas-dev/pandas/pandas/core/indexes/multi.py Imports are incorrectly sorted.
ERROR: /home/travis/build/pandas-dev/pandas/pandas/tests/indexes/multi/test_constructor.py Imports are incorrectly sorted.
Check import format using isort DONE
```

https://travis-ci.org/pandas-dev/pandas/jobs/460592619#L2916

```
Check that the deprecated `assert_raises_regex` is not used (`pytest.raises(match=pattern)` should be used instead)
pandas/tests/indexes/multi/test_constructor.py:    with tm.assert_raises_regex(TypeError, 'Input must be a DataFrame'):
pandas/tests/indexes/multi/test_constructor.py:        with tm.assert_raises_regex(TypeError, "'names' must be a list / "
Check that the deprecated `assert_raises_regex` is not used (`pytest.raises(match=pattern)` should be used instead) DONE
```

TomAugspurger commented:
Will need to merge master and fix the conflict as well.

TomAugspurger commented on pandas/tests/indexes/multi/test_constructor.py:
```diff
@@ -463,20 +492,71 @@ def test_create_index_existing_name(idx):
+])
+def test_from_frame_non_frame(non_frame):
+    # GH 22420
+    with tm.assert_raises_regex(TypeError, 'Input must be a DataFrame'):
```
This should be `with pytest.raises(TypeError, match='Input must be a DataFrame')` now.

ArtinSarraf replied:
fixed, thanks.

TomAugspurger commented on pandas/tests/indexes/multi/test_constructor.py:
```diff
@@ -463,20 +492,71 @@ def test_create_index_existing_name(idx):
+                      columns=pd.MultiIndex.from_tuples([('L1', 'x'),
+                                                         ('L2', 'y')]))
+    if names_out is None:
+        with tm.assert_raises_regex(TypeError, "'names' must be a list / "
```
Change to `pytest.raises`.

toobaz commented on pandas/core/indexes/multi.py:
```diff
@@ -1435,6 +1447,89 @@ def from_product(cls, iterables, sortorder=None, names=None):
+        elif is_list_like(names):
+            if len(names) != len(df.columns):
+                    raise ValueError("'names' should have same length as "
+                                     "number of columns in df.")
```
I suggested in my [previous comment](https://github.com/pandas-dev/pandas/pull/23141#discussion_r234456258) that all checks on ``names`` are superfluous: they exactly repeat the checks done inside the constructor. Isn't
```
names = column_names if names is None else names
```
sufficient?!

ArtinSarraf replied:
Yup, removed the redundant code. Thanks.

Commit:
CLN - clean up code and use pytest.raises

Commit:
Merge branch 'master' into enhancement/from_frame

Commit:
DOC - move to_frame breaking changes to backwards incompatible section of whatsnew

mitstake committed:
Merge branch 'master' into enhancement/from_frame

Commit:
Merge branch 'master' into enhancement/from_frame

Commit:
Merge branch 'enhancement/from_frame' of https://github.com/ArtinSarraf/pandas into enhancement/from_frame

jreback commented on pandas/tests/indexes/multi/test_constructor.py:
```diff
@@ -461,20 +490,72 @@ def test_create_index_existing_name(idx):
+@pytest.mark.parametrize('names_in,names_out', [
+    (None, [('L1', 'x'), ('L2', 'y')]),
+    (['x', 'y'], ['x', 'y']),
+    ('bad_input', ValueError("Names should be list-like for a MultiIndex")),
```
why don't you split this to 2 tests, with 1 the working cases, and 1 the error cases, easier to read

ArtinSarraf replied:
done.

jreback reviewed (changes requested):
some doc comments. can you update advanced.rst (see where we use .from_arrays and add an example using .from_frame). ping on green.

jreback commented:
looks pretty good to me, just some doc comments remain. @toobaz @datapythonista can you have a look and comment or approve.

Commit:
DOC - add advanced.rst section

Commit:
Merge branch 'master' into enhancement/from_frame

Commit:
Merge branch 'master' into enhancement/from_frame

jreback reviewed (changes requested):
small doc-comments. ping on green.

ArtinSarraf commented on pandas/tests/indexes/multi/test_constructor.py:
```diff
@@ -470,20 +499,79 @@ def test_create_index_existing_name(idx):
+    np.array([[1, 2], [3, 4], [5, 6]]),
+    27
+])
+def test_from_frame_non_frame(non_frame):
```
done.

ArtinSarraf replied:
done.

ArtinSarraf commented on pandas/tests/indexes/multi/test_conversion.py:
```diff
@@ -83,6 +85,39 @@ def test_to_frame():
+    # GH 22420
+    mi = pd.MultiIndex.from_arrays([
+        pd.date_range('19910905', periods=6, tz='US/Eastern'),
+        [1, 1, 1, 2, 2, 2],
```
This test was at the suggestion of @TomAugspurger 
https://github.com/pandas-dev/pandas/pull/23141#discussion_r230379498

Commit:
DOC/CLN - cleanup documentation

ArtinSarraf commented on doc/source/advanced.rst:
```diff
@@ -89,6 +90,17 @@ to use the :meth:`MultiIndex.from_product` method:
    iterables = [['bar', 'baz', 'foo', 'qux'], ['one', 'two']]
    pd.MultiIndex.from_product(iterables, names=['first', 'second'])
 
+You can also construct a ``MultiIndex`` from a ``DataFrame`` directly, using 
```
done.

ArtinSarraf commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -378,6 +378,7 @@ Backwards incompatible API changes
 - Passing scalar values to :class:`DatetimeIndex` or :class:`TimedeltaIndex` will now raise ``TypeError`` instead of ``ValueError`` (:issue:`23539`)
 - ``max_rows`` and ``max_cols`` parameters removed from :class:`HTMLFormatter` since truncation is handled by :class:`DataFrameFormatter` (:issue:`23818`)
 - :meth:`read_csv` will now raise a ``ValueError`` if a column with missing values is declared as having dtype ``bool`` (:issue:`20591`)
+- The column order of the resultant ``DataFrame`` from ``MultiIndex.to_frame()`` is now guaranteed to match the ``MultiIndex.names`` order. (:issue:`22420`)
```
done.

ArtinSarraf commented on pandas/core/indexes/multi.py:
```diff
@@ -431,6 +437,78 @@ def from_product(cls, iterables, sortorder=None, names=None):
+        3        1  jolly
+        4        2    joy
+        5        2    joy
+        >>> pd.MultiIndex.from_frame(df)
```
done.

ArtinSarraf commented on pandas/core/indexes/multi.py:
```diff
@@ -431,6 +437,78 @@ def from_product(cls, iterables, sortorder=None, names=None):
+                   labels=[[0, 0, 1, 1, 2, 2], [0, 1, 0, 1, 2, 2]],
+                   names=['will_be', 'used'])
+
+        >>> df = pd.DataFrame([['ahc', 'iam'], ['ahc', 'wim'], ['boh', 'amg'],
```
done.

Commit:
CLN - fix linting error according to pandas-dev.pandas test

ArtinSarraf commented:
@jreback -  any reason why I would get this linting error in pandas-dev.pandas tests ?

    Bash exited with code '1'.

Nothing else accompanying it (I fixed the other ones that were there before). The local pep8 checks don't show anything, and the branch is up to date with master.

Same thing is happening on this PR (https://github.com/pandas-dev/pandas/pull/23538)

datapythonista commented:
To see the error you need to click into the log. For many of them it's easy to make them red, and they are highlighted, but in this case it's an "error" with the ordering of the imports:
```
2018-12-06T03:44:33.5517264Z ERROR: /home/vsts/work/1/s/pandas/core/indexes/multi.py Imports are incorrectly sorted.
2018-12-06T03:44:33.5518052Z ERROR: /home/vsts/work/1/s/pandas/tests/indexes/multi/test_constructor.py Imports are incorrectly sorted.
```

datapythonista commented on pandas/core/indexes/multi.py:
```diff
@@ -431,6 +437,70 @@ def from_product(cls, iterables, sortorder=None, names=None):
+
+        Parameters
+        ----------
+        df : pd.DataFrame
```
```suggestion
        df : DataFrame
```

ArtinSarraf replied:
fixed

datapythonista commented on pandas/core/indexes/multi.py:
```diff
@@ -431,6 +437,70 @@ def from_product(cls, iterables, sortorder=None, names=None):
+        ----------
+        df : pd.DataFrame
+            DataFrame to be converted to MultiIndex.
+        sortorder : int or None
```
```suggestion
        sortorder : int, optional
```

And please explain what it means to not be provided (`None`)

ArtinSarraf replied:
done

datapythonista commented on pandas/core/indexes/multi.py:
```diff
@@ -431,6 +437,70 @@ def from_product(cls, iterables, sortorder=None, names=None):
+        Returns
+        -------
+        MultiIndex or Index
+            The MultiIndex representation of the given DataFrame.
```
I guess it'll be `Index` when the `DataFrame` has a single column. Would be useful to explain that.

ArtinSarraf replied:
fixed, we changed this so that it's always MultiIndex now

datapythonista commented on pandas/core/indexes/multi.py:
```diff
@@ -431,6 +437,70 @@ def from_product(cls, iterables, sortorder=None, names=None):
+        >>> df = pd.DataFrame([['ahc', 'happy'], ['ahc', 'jolly'],
+        ...                    ['boh', 'happy'], ['boh', 'jolly'],
+        ...                    ['oil', 'joy'], ['oil', 'joy']],
+        ...                   columns=['a', 'b'])
```
This is too long for what is needed. Also, much better to use some meaningful data. We've been using in many places a small dataset of animals and their number of legs, wings... I'd use the same here.

ArtinSarraf replied:
Changed. I didn't want to use a dataset with numeric data since I think that the example should demonstrate the tie between multiindexes and metadataframes. Instead I used a weather related example (which was the inspiration for this addition in the first place)

datapythonista commented on pandas/core/indexes/multi.py:
```diff
@@ -431,6 +437,70 @@ def from_product(cls, iterables, sortorder=None, names=None):
+                   labels=[[0, 0, 1, 1, 2, 2], [0, 1, 0, 1, 2, 2]],
+                   names=['a', 'b'])
+
+        # Use explicit names, instead of column names
```
Better as a paragraph than as a code comment (remove the `#` and leave a blank line after this line.

ArtinSarraf replied:
done.

datapythonista commented on pandas/core/indexes/multi.py:
```diff
@@ -431,6 +437,70 @@ def from_product(cls, iterables, sortorder=None, names=None):
+                   labels=[[0, 0, 1, 1, 2, 2], [0, 1, 0, 1, 2, 2]],
+                   names=['X', 'Y'])
+
+        See Also
```
This goes before the Examples section. `./scripts/validate_docstrings.py pandas.MultiIndex.from_frame` should report it as an error. Make sure nothing is reported by the script.

ArtinSarraf replied:
Fixed this docstring and all the other constructor methods docstrings (since I had to modify them to update the See Alsos). I also fixed the pd.MultiIndex docstring to the best of my abilities (since I had to make some small modifications to that too). However, there were still some issues with the MultiIndex docstring:

    2 Errors found:
        Parameters {_set_identity, name, dtype} not documented
        Unknown parameters {labels}

Any ideas on how I should address these? Looks like labels is there to serve as a deprecation reminder.

datapythonista reviewed (changes requested):
Looks good, the docstring has some format errors and can be improved, but good work.

Commit:
DOC - fix docstrings

Commit:
CLN - fix import order with isort

ArtinSarraf commented:
@ jreback - tests are clean.

jreback commented:
@ArtinSarraf one more merge master and looks good.

jreback commented:
@datapythonista @toobaz if you have further comments.

datapythonista reviewed (approved):
lgtm, thanks for all the fixes to docstrings @ArtinSarraf

Commit:
Merge branch 'master' into enhancement/from_frame

ArtinSarraf commented:
@jreback - master has been merged and tests are green.

toobaz commented:
Looks good to me!

jreback commented:
thanks @ArtinSarraf
